### DISCUSSION TITLE
Subquery update not requested

### DISCUSSION THREAD

**Discussion by marco-mastella** [prisma/prisma#23771] [Q&A]
### Question

Ciao,
i have the schema as mentioned in the box below
considering I have 1 subscription linked to 1 subscriptionPlan. currentSubscriptionPlanId is filled with subscriptionPlan.id
now I need to change the subscriptionPlan (keeping the history) so what I'm trying to do is:
- create a new SubscriptionPlan
- updated old SubscriptionPlan status to "archived"
- update subscription.currentSubscriptionPlanId with the id of the new subscriptionPlan.

when I try to do this in a single transaction, I receive an error
{"name":"PrismaClientKnownRequestError","code":"P2002","clientVersion":"5.12.1","meta":{"modelName":"Subscription","target":["id"]}}

I tried to do all the query without transaction and what I see is that prisma when I'm updating Subscription table is also updating subscriptionPlan table.
`prisma:query UPDATE "public"."SubscriptionPlan" SET "id" = $1 WHERE ("public"."SubscriptionPlan"."id" IN ($2) AND 1=1)`

`prisma:query UPDATE "public"."Subscription" SET "currentSubscriptionPlanId" = $1, "idUserUpd" = $2, "updTS" = $3 WHERE ("public"."Subscription"."id" = $4 AND 1=1) RETURNING "public"."Subscription"."id", "public"."Subscription"."idCompany",  "public"."Subscription"."currentSubscriptionPlanId", ..."public"."Subscription"."status", "public"."Subscription"."insTS", "public"."Subscription"."updTS", "public"."Subscription"."idUserIns", "public"."Subscription"."idUserUpd"`

### Expected behavior (optional)

i would expect to update as requested

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
model Subscription {
  id                        String                @id @default(uuid())
  idCompany                 String?               @default("")
  currentSubscriptionPlanId String?               @default("")
  ...
  status                    String?               @default("ACTIVE") //Status
  insTS                     DateTime?             @default(now())
  updTS                     DateTime?             @updatedAt
  idUserIns                 String?               @default("")
  idUserUpd                 String?               @default("")
  company                   Company[]
  subscriptionPlan          SubscriptionPlan[]    @relation("subscriptionPlan")
  currentSubscriptionPlan   SubscriptionPlan?     @relation("currentSubscriptionPlan")

  @@unique([currentSubscriptionPlanId, idCompany])
  @@index([idCompany, status])
}

model SubscriptionPlan {
  id                    String                @id @default(uuid())
  idCompany             String?               @default("")
  idSubscription        String?               @default("")
   ...
  status                String?               @default("ACTIVE") //Status
  insTS                 DateTime?             @default(now())
  updTS                 DateTime?             @updatedAt
  idUserIns             String?               @default("")
  idUserUpd             String?               @default("")
  subscription          Subscription?         @relation("subscriptionPlan", fields: [idSubscription], references: [id])
  currentSubscription   Subscription?         @relation("currentSubscriptionPlan", fields: [id, idCompany], references: 
 [currentSubscriptionPlanId, idCompany])

  @@unique([id, idCompany])
  @@index([idCompany, status])
  @@index([idSubscription, status])

}
```

```ts
await dbClient.$transaction([
            dbClient.subscriptionPlan.create({
                data: newSubscriptionPlan,
            }),
            dbClient.subscriptionPlan.update({
                where: {
                    id: oldSubscriptionPlan.id,
                    idSubscription: oldSubscription.id,
                    idCompany: companyId,
                },
                data: {
                    status: DBStatus.ARCHIVED,
                    idUserUpd: userId,
                    updTS: new Date(),
                },
            }),
            dbClient.subscription.update({
                where: {
                    id: oldSubscription.id,
                    currentSubscriptionPlanId: oldSubscriptionPlan.id,
                    idCompany: companyId,
                },
                data: {
                    currentSubscriptionPlanId: newIDSubPlan,
                    idUserUpd: userId,
                    updTS: new Date(),
                },
            }),
```

- OS: firebase functions
- Database: PostgreSQL
- Node.js version: node 20

- Run `prisma -v` to see your Prisma version and paste it
prisma                  : 5.12.1
@prisma/client          : 5.12.1
Computed binaryTarget   : windows
Operating System        : win32
Architecture            : x64
Node.js                 : v20.10.0
Query Engine (Node-API) : libquery-engine 473ed3124229e22d881cb7addf559799debae1ab (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Schema Engine           : schema-engine-cli 473ed3124229e22d881cb7addf559799debae1ab (at node_modules\@prisma\engines\schema-engine-windows.exe)
Schema Wasm             : @prisma/prisma-schema-wasm 5.12.0-21.473ed3124229e22d881cb7addf559799debae1ab
Default Engines Hash    : 473ed3124229e22d881cb7addf559799debae1ab
Studio                  : 0.499.0

**Comment by ludralph** [comment]
Hi @marco-mastella 👋 

Thank you for raising this question.

It seems the update on the `Subscription` table tries to set `currentSubscriptionPlanId` to the new plan's id, but this might violate the unique constraint on (currentSubscriptionPlanId, idCompany). Can you rewrite your code as shown below?
```js
const newSubscriptionPlan = await dbClient.subscriptionPlan.create({
  data: newSubscriptionPlanData,
});

await dbClient.$transaction([
  dbClient.subscription.update({
    where: {
      id: oldSubscription.id,
      currentSubscriptionPlanId: oldSubscriptionPlan.id,
      idCompany: companyId,
    },
    data: {
      currentSubscriptionPlanId: newSubscriptionPlan.id,
      idUserUpd: userId,
      updTS: new Date(),
    },
  }),
  dbClient.subscriptionPlan.update({
    where: {
      id: oldSubscriptionPlan.id,
      idSubscription: oldSubscription.id,
      idCompany: companyId,
    },
    data: {
      status: DBStatus.ARCHIVED,
      idUserUpd: userId,
      updTS: new Date(),
    },
  }),
]);
```
We first create the new `SubscriptionPlan` outside the transaction.  Inside the transaction, we:

- Update the Subscription's `currentSubscriptionPlanId` with the new plan's id.
- Update the old SubscriptionPlan's status to "ARCHIVED".

**Reply by marco-mastella** [reply]
ciao, it doesn't work.

i receive the following error
`data: "{"name":"PrismaClientKnownRequestError","code":"P2002","clientVersion":"5.12.1","meta":{"modelName":"Subscription","target":["id"]}}"`

i can understand your requirement for an index, but I strongly recommend removing the requirement of a **UNIQUE** index.

this is the code
```
const newSubPlanRes = await dbClient.subscriptionPlan.create({
            data: newSubscriptionPlan,
        })

        const [archiveRes, subRes] = await dbClient.$transaction([
            dbClient.subscriptionPlan.update({
                where: {
                    id: oldSubscriptionPlan.id,
                    idSubscription: oldSubscription.id,
                    idCompany: companyId,
                },
                data: archiveSubPlan,
            }),

            dbClient.subscription.update({
                where: {
                    id: oldSubscription.id,                   
                    idCompany: companyId,
                },
                data: {
                    currentSubscriptionPlanId: newSubscriptionPlan.id,
                    idUserUpd: userId,
                    updTS: new Date(),
                },
            }),
        ])
```
please consider my current Prisma config
```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider     = "postgresql"
  url          = env("DATABASE_URL")
  relationMode = "prisma"
}
```

![image](https://github.com/prisma/prisma/assets/78215504/53142334-04d7-41e7-880b-c92833c45dd7)

**Reply by ludralph** [reply]
Hi @marco-mastella 

Has this been resolved? Can you try this out?
```js
const newSubPlanRes = await dbClient.subscriptionPlan.create({
  data: newSubscriptionPlan,
});

await dbClient.$transaction(async (tx) => {
  // First, archive the old subscription plan
  await tx.subscriptionPlan.update({
    where: {
      id: oldSubscriptionPlan.id,
      idSubscription: oldSubscription.id,
      idCompany: companyId,
    },
    data: archiveSubPlan,
  });

  // Then, update the subscription
  await tx.subscription.update({
    where: {
      id: oldSubscription.id,
      idCompany: companyId,
    },
    data: {
      currentSubscriptionPlanId: newSubPlanRes.id,
      idUserUpd: userId,
      updTS: new Date(),
    },
  });
});
```
With this code, we are separating the creation of the new `SubscriptionPlan` from the transaction and performs the updates sequentially within the transaction.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;
